libportal/settings.c

```diff
@@ -0,0 +1,191 @@
+
+  if (settings->signal_id)
+    g_dbus_connection_signal_unsubscribe (settings->portal->bus, settings->signal_id);
+
+  g_clear_object (&settings->portal);
+
+  G_OBJECT_CLASS (xdp_settings_parent_class)->finalize (object);
+}
+
+static void
+xdp_settings_class_init (XdpSettingsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = xdp_settings_finalize;
+
+  /**
+   * XdpSettings::changed:
+   * @settings: the [class@Settings] object
+   * @namespace: the value namespace
+   * @key: the value key
+   * @value: the value
+   *
+   * Emitted when a setting value is changed externally.
+   */
+  signals[CHANGED] =
+    g_signal_new ("changed",
+                  G_TYPE_FROM_CLASS (object_class),
+                  G_SIGNAL_RUN_FIRST,
+                  0,
+                  NULL, NULL,
+                  NULL,
+                  G_TYPE_NONE, 3,
+                  G_TYPE_STRING,
+                  G_TYPE_STRING,
+                  G_TYPE_VARIANT);
+}
+
+static void
+xdp_settings_init (XdpSettings *settings)
+{
+}
+
+static void
+settings_changed (GDBusConnection *bus,
+		  const char *sender_name,
+		  const char *object_path,
+		  const char *interface_name,
+		  const char *signal_name,
+		  GVariant *parameters,
+		  gpointer data)
+{
+  const char *namespace = NULL;
+  const char *key = NULL;
+  g_autoptr(GVariant) value = NULL;
+  guint n_params;
+
+  XdpSettings *settings = data;
+
+  n_params = g_variant_n_children (parameters);
+  if (n_params != 3)
+    {
+      g_warning ("Incorrect number of parameters, expected 3, got %u", n_params);
+      return;
+    }
+
+  g_variant_get_child (parameters, 0, "&s", &namespace);
+  g_variant_get_child (parameters, 1, "&s", &key);
+  g_variant_get_child (parameters, 2, "v", &value);
+
+  g_signal_emit (settings, signals[CHANGED], 0, namespace, key, value);
+}
+
+/**
+ * xdp_settings_read:
+ * @settings: the [class@Settings] object.
+ * @namespace: the namespace of the value.
+ * @key: the key of the value.
+ *
+ * Read a setting value within @namespace, with @key.
+ *
+ * Returns: (nullable) (transfer full): the value, or %NULL if not found.
+ */
+XDP_PUBLIC
+GVariant *xdp_settings_read (XdpSettings *settings, const char *namespace, const char *key)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) ret = NULL;
+  g_autoptr(GVariant) inner = NULL;
+
+  ret = g_dbus_connection_call_sync (settings->portal->bus,
+                                     PORTAL_BUS_NAME,
+                                     PORTAL_OBJECT_PATH,
+                                     SETTINGS_INTERFACE,
+                                     "ReadOne",
+                                     g_variant_new ("(ss)", namespace, key),
+                                     NULL,
+                                     G_DBUS_CALL_FLAGS_NONE,
+                                     -1,
+                                     NULL, &error);
```

Comment:
Let's expose `error` as an input so callers understand why it fails.

Reply:
done